firedrake/randomfunctiongen.py

```diff
@@ -235,103 +215,169 @@ def func(self, *args, **kwargs):
+    elif module_attr in ['MT19937', 'Philox', 'PCG64', 'SFC64']:
         _Base = getattr(randomgen, module_attr)
+        __doc__ = _reformat_doc(getattr(randomgen, module_attr).__doc__)
 
         def __init__(self, *args, **kwargs):
-            self._comm = kwargs.pop('comm', MPI.COMM_WORLD)
-            # Remember args, kwargs as these are changed in super().__init__
-            _args = args
-            _kwargs = kwargs
-            self._initialized = False
-            super(_Wrapper, self).__init__(*args, **kwargs)
-            self._initialized = True
-            self.seed(*_args, **_kwargs)
-            self._generator = None
-
-        @property
-        def generator(self):
-            if self._generator is None:
-                from . import RandomGenerator
-                self._generator = RandomGenerator(brng=self)
-            return self._generator
+            _kwargs = kwargs.copy()
+            self._comm = _kwargs.pop('comm', MPI.COMM_WORLD)
+            if self._comm.Get_size() > 1 and module_attr not in ['PCG64', 'Philox']:
+                raise TypeError("Use 'PCG64', 'Philox', for parallel RNG")
+            self._init(*args, **_kwargs)
 
-        if module_attr == 'PCG64':
+        def seed(self, *args, **kwargs):
+            raise AttributeError("`seed` method is not available in `numpy.random`; if reseeding, create a new bit generator with the new seed.")
 
+        if module_attr == 'PCG64':
             # Use examples in https://bashtage.github.io/randomgen/parallel.html
             # with appropriate changes.
-            def seed(self, seed=None, inc=None):
-
-                if self._comm.Get_size() == 1:
-                    super(_Wrapper, self).seed(seed=seed, inc=0 if inc is None else inc)
-                else:
-                    rank = self._comm.Get_rank()
-                    if seed is None:
-                        if rank == 0:
-                            # generate a 128bit seed
-                            entropy = randomgen.entropy.random_entropy(4)
-                            seed = sum([int(entropy[i]) * 2 ** (32 * i) for i in range(4)])
-                        else:
-                            seed = None
-                        # All processes have to have the same seed
-                        seed = self._comm.bcast(seed, root=0)
-                    # Use rank to generate multiple streams.
-                    # If 'inc' is to be passed, it is users' responsibility
-                    # to provide an appropriate value.
-                    inc = inc or rank
-                    super(_Wrapper, self).seed(seed=seed, inc=inc)
-
-        elif module_attr in ('Philox', 'ThreeFry'):
-
-            def seed(self, seed=None, counter=None, key=None):
-
+            def _init(self, *args, **kwargs):
+                # numpy.random does not allow users to set `seed` and `inc` separately,
+                # while randomgen (v.1.19) does. We follow the randomgen interface here.
```

Comment:
Given that randomgen is deprecated and numpy is the stable API, can we not do this?

Reply:
So `randomgen` stopped maintaining `randomgen.Generator`, and we are now recommended to use `numpy.random.Generator`, but it turned out that `numpy.random` is not a straightforward wrapper of `randomgen`. As far as I understand, one cool thing about `PCG64` is that one can create parallel-safe streams by setting `inc` parameter to different values while keeping the `seed` the same, and this is the recommended way to create multiple streams in parallel, and, indeed, `randomgen.PCG64` (latest version) still takes both `seed` and `inc` parameters. `numpy.random.PCG64`, however, somehow only takes one argument (`seed`), and generates two parameters from that `seed`, and pass those to `randomgen.PCG64` constructor under the hood. I don't know if this would create parallel-safe streams when a user make the value of the `seed` `rank` dependent as I don't know much about bit generators, so I decided to follow `randomgen.PCG64` interface, so that `Firedrake.PCG64` still takes two arguments (`seed` and `inc`). Our interface is also backward compatible this way.

Reply:
Per https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html
> The preferred way to use a BitGenerator in parallel applications is to use the SeedSequence.spawn method to obtain entropy values, and to use these to generate new BitGenerators:

```python
from numpy.random import Generator, PCG64, SeedSequence
sg = SeedSequence(1234)
rg = [Generator(PCG64(s)) for s in sg.spawn(10)]
```

Reply:
Ok, I now made this default, but explained how to explicitly use `inc` in the module docstring. Thanks!